### DISCUSSION TITLE
[Feature] Add support for custom logout url/endpoint for OIDC authentication via Authentik

### DISCUSSION THREAD

**Discussion by abjoseph** [immich-app/immich#11104] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

When Immich is configured for OIDC authentication via Authentik, initiating a logout action redirects back to the Authentik "session-end". However, when Immich is configured for OIDC via Google, initiating a logout action simply redirects the user back to the Immich login screen. 

In most cases, when a user logs out of the service provider (Immich) the intention is not to end or logout of identity provider but simply to end the authenticated session in the current service (Immich). At worst, this is merely an inconsistency in behavior across two different IdPs but it would be great if the Immich administrator could configure a custom logout url, which in most cases, would simply be redirecting the user back to the Immich login screen and not having to be redirected to the IdP end session screen. I could easily see a scenario in which a non-technical user accidentally logs out of the IdP and then as to re-authenticate with password and TOTP/Passkey, etc.

1. In the first screenshot below, you can see that Immich successfully processes the logout action.
![image](https://github.com/user-attachments/assets/2d56a09f-789b-438d-b10e-4d93fffc22be)

2. However, in this second screenshot, you can see Immich then immediately peforms a 302 redirect to the IdP (Authentik) session-end endpoint e.g `https://auth.example.com/application/o/immich/end-session/`, which is non-configurable currently because Immich reads this information from the discovery endpoint.

![image](https://github.com/user-attachments/assets/af6dcaf5-46a4-4385-87d3-7d10f7bd957a)

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by abjoseph** [comment]
Bumping for visibility.

**Reply by bo0tzz** [reply]
Please don't bump unnecessarily, it adds nothing.

**Comment by MahmoudAlyuDeen** [comment]
Until an official solution becomes available from either immich or authentik, I'm configuring my reverse proxy to do a simple redirect to my desired destination.

In my case, I want to go back to authentik home page without tapping the blue button.

![Screenshot 2024-10-10 at 11 18 12 AM](https://github.com/user-attachments/assets/a1fcc0d2-ffdb-49db-98f7-9d2a618fcd87)

I'm using nginx proxy manager, but I imagine any reverse proxy solution can do a simple 302 redirect to a configured URL.

**Comment by sakibstark11** [comment]
Strangely for me, it is impossible to log out of immich, when I click on the log out button, it takes me to logging in page. However the same authentik config works as expected for my other services. Did you have similar issues?

**Reply by heissonwillen** [reply]
Do you have OAuth "AUTO LAUNCH" configured on Immich? 

I assume what happens is something like:
You logout from Immich &rarr; you get redirected to the login page on Immich &rarr; you get back logged in, as you were never logged out from the IdP, and Immich automatically starts the OAuth flow.

The work around in this case would be to disable the OAuth auto launch.

**Comment by heissonwillen** [comment]
I've managed to have this flow working using Keycloak as the IdP.

I implemented a solution based on the [OpenID RP-Initiated](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout) spec, which consisted of sending the IdP `id_token_hint` and `post_logout_redirect_uri` parameters back to the IdP on the logout request.

By doing this:
- The IdP could identify which client I was trying to log out from;
- The IdP redirected me back to the Immich login screen.

Here's what I mean:
https://github.com/user-attachments/assets/b835b051-1386-4174-859b-547131952486

Does this implementation + feature make sense to you, @bo0tzz? If so, I'd like to discuss a few topics on it:

**1. Where to store the IdP `id_token_hint`?**

I've initially had two ideas:
- A new cookie, something like `immich_idp_id_token` or `immich_oauth_id_token`; OR
- A new field on the `sessions` table on the DB - the preferred solution from my side, as the client wouldn't have the chance to mess up with the logout flow.

**2. What should be used as the `post_logout_redirect_uri`?**

I've experimented trying to extract the host from the express `request` object, but it didn't work. The requests had the internal docker network as the host (i.e.: `immich-server:2238`). I believe this issue would even be increased in cases that are more sophisticated than mine (e.g.: reverse-proxies, custom ingresses, and so on).

My proposal here would be to rely on the user to set this value, by adding a new field on the OAuth settings. The 'drawback' is that the OAuth logout wouldn't be fully supported in case you want to access Immich from multiple hostnames.

---

Once we agree upon these, I'll be more than happy to open up a PR implementing the solution. :)

**Comment by bo0tzz** [comment]
I'm trying to understand the issue/request here, but it's not entirely clear to me. As I understand it, the logout works as expected with Authentik but not with Google. If I compare those providers' openid discovery documents, Authentik provides an `end_session_endpoint` and Google doesn't, so it seems to me like that's pretty much working as expected and Google doesn't support the desired feature. I don't understand what Immich could do differently here, or how the proposed PR would resolve it (aiui, we're already using that PR-Initiated endpoint).

**Reply by heissonwillen** [reply]
> aiui, we're already using that PR-Initiated endpoint

Yep, you're right. Immich uses it, but it doesn't send any paramater to that endpoint - the parameters are used as a "clue" for the IdP to properly handle the logout flow. 

None of them is required by the spec, though - meaning that the Google IdP is free to implement it's own way to handle logout/properly redirect the user at the end of the login flow.

**Reply by heissonwillen** [reply]
Ahh, my bad. I've noticed I focused too much on https://github.com/immich-app/immich/discussions/11104#discussioncomment-10901945, instead of the original comment.

**Comment by peterfiabigron** [comment]
Hi, i am setting up authentik with immich to secure my valuable memories in this fantastic app. As far as i can understand now, the logout process is fixed, it will always take me back to a page, where i have to choose whether i want to also log out from authentik or not. I think this poses a relatively big security risk, since i am not convinced that my users will understand the consequence and just close the browser without making the second logout. If it happens on a not secure device it means that anyone will be able to access the data for a long time, especially if remember me function was also used. I think it would be much cleaner if logout could be configured in a way that avoids this user error. Is there any chance this will be added to immich features?

**Reply by peterfiabigron** [reply]
As i dig deeper in, i assume that there is more to that. Oidc in immich seems to be used only for login. So sessions duration, remember device, stay logged in offset and logout behaviour are not controlled by IdP/authentik, but those are internal immich hardcoded properties. Is there any chance it will change and immich will give full control to IdPs for these behaviours? It would make sense, in my opinion, to let the whole access security be handled by IdPs to make it safe and customizable. What do you think?

**Reply by waltech25** [reply]
Same thing I observed. I use google auth with pangolin and immich as well. So i just sign in once with google in pangolin and it automatically signs into immich with my account. But if i sign out from immich and reload the immich url. I can sign back in with just one click. Which should not be.

If we can add the option for the users to add the custom logout url behaviour and then we can add the url to execute : 
accounts.google.com/logout 
It will then logout the google session completely. 

It could be optional how the users want the logout behaviour and they can call multiple urls for logout.

**Reply by ovizii** [reply]
Same problem here with Pocket-ID as my IDP.

**Reply by daniel-l** [reply]
> Same problem here with Pocket-ID as my IDP.

Same here with Voidauth.

**Comment by antonio-ivanovski** [comment]
Facing the same issue with Auth0, need to manually navigte to `https://myauth.myapp.com/logout` to cleanup the Auth0 Session. From what I can see Auth0 doesn't provide the `end_session_endpoint` as part of the well-known config.

Am I doing something wrong?

**Comment by bo0tzz** [comment]
#27389
